Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,804 @@
+  for (auto& s : skylanders)
+  {
+    if (s.status & 1)
+    {
+      s.queued_status.push(Skylander::ADDED);
+      s.queued_status.push(Skylander::READY);
+    }
+  }
+
+  m_activated = true;
+}
+
+void SkylanderPortal::Deactivate()
+{
+  std::lock_guard lock(sky_mutex);
+
+  for (auto& s : skylanders)
+  {
+    // check if at the end of the updates there would be a figure on the portal
+    if (!s.queued_status.empty())
+    {
+      s.status = s.queued_status.back();
+      s.queued_status = std::queue<u8>();
+    }
+
+    s.status &= 1;
+  }
+
+  m_activated = false;
+}
+
+bool SkylanderPortal::IsActivated()
+{
+  std::lock_guard lock(sky_mutex);
+
+  return m_activated;
+}
+
+void SkylanderPortal::UpdateStatus()
+{
+  std::lock_guard lock(sky_mutex);
+
+  if (!m_status_updated)
+  {
+    for (auto& s : skylanders)
+    {
+      if (s.status & 1)
+      {
+        s.queued_status.push(Skylander::REMOVED);
+        s.queued_status.push(Skylander::ADDED);
+        s.queued_status.push(Skylander::READY);
+      }
+    }
+    m_status_updated = true;
+  }
+}
+
+// Side:
+// 0x00 = right
+// 0x01 = left and right
+// 0x02 = left
+// 0x03 = trap
+void SkylanderPortal::SetLEDs(u8 side, u8 red, u8 green, u8 blue)
+{
+  std::lock_guard lock(sky_mutex);
+  if (side == 0x00)
+  {
+    m_color_right.red = red;
+    m_color_right.green = green;
+    m_color_right.blue = blue;
+  }
+  else if (side == 0x01)
+  {
+    m_color_right.red = red;
+    m_color_right.green = green;
+    m_color_right.blue = blue;
+
+    m_color_left.red = red;
+    m_color_left.green = green;
+    m_color_left.blue = blue;
+  }
+  else if (side == 0x02)
+  {
+    m_color_left.red = red;
+    m_color_left.green = green;
+    m_color_left.blue = blue;
+  }
+  else if (side == 0x03)
+  {
+    m_color_trap.red = red;
+    m_color_trap.green = green;
+    m_color_trap.blue = blue;
+  }
+}
+
+std::array<u8, 64> SkylanderPortal::GetStatus()
+{
+  std::lock_guard lock(sky_mutex);
+
+  u32 status = 0;
+  u8 active = 0x00;
+
+  if (m_activated)
+  {
+    active = 0x01;
+  }
+
+  for (int i = MAX_SKYLANDERS - 1; i >= 0; i--)
+  {
+    auto& s = skylanders[i];
+
+    if (!s.queued_status.empty())
+    {
+      s.status = s.queued_status.front();
+      s.queued_status.pop();
+    }
+    status <<= 2;
+    status |= s.status;
+  }
+
+  std::array<u8, 64> result = {0x53,   0x00, 0x00, 0x00, 0x00, m_interrupt_counter++,
+                               active, 0x00, 0x00, 0x00, 0x00, 0x00,
+                               0x00,   0x00, 0x00, 0x00, 0x00, 0x00,
+                               0x00,   0x00, 0x00, 0x00, 0x00, 0x00,
+                               0x00,   0x00, 0x00, 0x00, 0x00, 0x00,
+                               0x00,   0x00};
+  memcpy(&result[1], &status, sizeof(status));
+  return result;
+}
+
+void SkylanderPortal::QueryBlock(u8 sky_num, u8 block, u8* reply_buf)
+{
+  std::lock_guard lock(sky_mutex);
+
+  const auto& skylander = skylanders[sky_num];
```

Comment:
Might be worth adding a bounds check to make sure `sky_num` is a valid index.

```cpp
if (sky_num >= skylanders.size())
  return;
```

And `block` should probably also be checked.

As WriteBlock will contain the exact same check, it may be worth putting the validation logic in a separate "is valid sky num and block" function that can be used from both QueryBlock and WriteBlock